src/NServiceBus.Gateway/Gateway.cs

```diff
@@ -53,19 +46,42 @@ protected override void Setup(FeatureConfigurationContext context)
 
             var gatewayPipeline = context.AddSatellitePipeline("Gateway", "gateway", requiredTransactionSupport, PushRuntimeSettings.Default, out gatewayInputAddress);
 
-            ConfigureChannels(context);
+            var channelManager = CreateChannelManager(context);
+            var channelFactory = CreateChannelFactory(context);
+            
+            gatewayPipeline.Register("GatewaySendProcessor", b => new GatewaySendBehavior(gatewayInputAddress, channelManager, new MessageNotifier(), b.Build<IDispatchMessages>(), context.Settings, CreateForwarder(context, channelFactory, b.BuildAll<IDataBus>()?.FirstOrDefault()), CreateSiteRouters(context)), "Processes messages to be sent to the gateway");
+            context.Pipeline.Register("RouteToGateway", b => new RouteToGatewayBehaviour(gatewayInputAddress), "Reroutes gateway messages to the gateway");
 
-            ConfigureReceiver(context);
+            RegisterHttpListenerInstaller(context);
             context.Pipeline.Register("GatewayIncomingBehavior", typeof(GatewayIncomingBehavior), "Extracts gateway related information from the incoming message");
             context.Pipeline.Register("GatewayOutgoingBehavior", typeof(GatewayOutgoingBehavior), "Puts gateway related information on the headers of outgoing messages");
+            context.RegisterStartupTask(b => new GatewayReceiverStartupTask(channelManager, channelFactory, GetEndpointRouter(context), b.Build<IDispatchMessages>(), b.Build<IDeduplicateMessages>(), b.BuildAll<IDataBus>()?.FirstOrDefault(), gatewayInputAddress));
```

Comment:
Seriously dislike `b.BuildAll<IDataBus>()?.FirstOrDefault()`, but this is the only way I could get an optional dependency to work. `IBuilder.Build<T>` blows up if no component has been registered and does not provide any methods to check for registration up front. Are we missing something in the `IBuilder` interface and/or a parameter for the `Register`/`RegisterStartupTask` methods? This will pop up as we remove IoC elsewhere.

Better way to do this @danielmarbach?